lib/cn/cn_tree.c

```diff
@@ -1855,49 +1919,170 @@ cn_comp_cleanup(struct cn_compaction_work *w)
 
-    if (ev(tree->ct_rspills_wedged && !w->cw_err)) {
-        w->cw_err = merr(ESHUTDOWN);
-        w->cw_canceled = true;
-    }
+    blk_list_free(&ss->ss_mblks.kblks);
+    blk_list_free(&ss->ss_mblks.vblks);
 
 done:
-    mutex_unlock(&tree->ct_rspills_lock);
+    w->cw_t4_commit = get_time_ns();
+
+    if (err) {
+        cndb_record_nak(cndb, tx);
+        kvset_put_ref(kvset);
+    }
+
+    return err;
+}
+
+static merr_t
+cn_node_apply_subspill(struct subspill *ss)
+{
+    struct cn_compaction_work *w = ss->ss_work;
+    merr_t err;
+
+    err = cn_subspill_commit(ss);
+    if (err)
+        return err;
+
+    if (ss->ss_added) {
+        w->cw_output_nodev[0] = ss->ss_node;
+        w->cw_checkpoint(w);
+    }
 
-    return w;
+    return 0;
+}
+
+static merr_t
+cn_comp_spill(struct cn_compaction_work *w)
+{
+    struct subspill *ss;
+    struct spillctx *sctx;
+    struct cn_tree *tree = w->cw_tree;
+    struct route_node *rtn = 0;
+    uint cnum = 0;
+    merr_t err;
+
+    err = cn_spill_init(w, &sctx);
+    if (err)
+        return err;
+
+    while (!err) {
+        struct cn_tree_node *node;
+        void *lock;
+        struct kvset_list_entry *le;
+        uint64_t node_dgen;
+        unsigned char ekey[HSE_KVS_KEY_LEN_MAX];
+        uint eklen;
+
+        rmlock_rlock(&tree->ct_lock, &lock);
+
+        rtn = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
+        if (!rtn) {
+            rmlock_runlock(lock);
+            break;
+        }
+
+        node = route_node_tnode(rtn);
+
+        le = list_first_entry_or_null(&node->tn_kvset_list, typeof(*le), le_link);
+        node_dgen = le ? kvset_get_dgen(le->le_kvset) : 0;
+
+        route_node_keycpy(rtn, ekey, sizeof(ekey), &eklen);
+        rmlock_runlock(lock);
+
+        ss = cn_spill_get_nth_subspill(sctx, cnum);
+
+        err = cn_subspill(sctx, ss, node, node_dgen, ekey, eklen);
+        if (err)
+            break;
+
+        /* Enqueue subspill only if there are older spills that need to update this node
+         * before us.
+         */
+        if (ss->ss_sgen == atomic_read(&node->tn_sgen) + 1) {
+            err = cn_node_apply_subspill(ss);
+            if (err)
+                break;
+
+            atomic_inc(&node->tn_sgen);
+        } else {
+            cn_node_enqueue_subspill(ss, node);
+        }
+
+        /* Apply subspills that are ready. */
+        while ((ss = cn_node_first_subspill(node))) {
+            err = cn_node_apply_subspill(ss);
+            if (err)
+                break;
+
+            atomic_inc(&node->tn_sgen);
+        }
+
+        ++cnum;
+    }
+
+    cn_spill_fini(sctx);
+    w->cw_t3_build = get_time_ns();
+
+    if (err && merr_errno(err) != ESHUTDOWN)
+        kvdb_health_error(tree->ct_kvdb_health, err);
+
+    /* Serialize the deletion of input kvsets.
+     */
+    if (!err && cn_node_spill_wait(w, w->cw_tree->ct_root))
+        cn_input_kvsets_delete(w);
+
+    return err;
```

Comment:
Cool, I see you added removal of stranded subspill records in your wip branch should the thread leave the function due to an async error.  In your cnum loop, (i <= cnum) seems wrong if the thread is cleaning up due to an async error condition rather than an error it encountered in the node walking loop, right?